File: java/money/ch14/src/main/java/com/thesoftwaregorilla/tdd/money/CurrencyPair.java
package com.thesoftwaregorilla.tdd.money;

import java.util.Objects;

public class CurrencyPair {
    private final String from;
    private final String to;

    public CurrencyPair(String from, String to) {
        this.from = from;
        this.to = to;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof CurrencyPair)) return false;
        CurrencyPair pair = (CurrencyPair) obj;
        return from.equals(pair.from) && to.equals(pair.to);
    }

    @Override
    public int hashCode() {
        return Objects.hash(from, to);
    }
}

File: cs/ch14/Money/CurrencyPair.cs
namespace TheSoftwareGorilla.TDD.Money;

public class CurrencyPair
{
    private readonly string _from;
    private readonly string _to;

    public CurrencyPair(string from, string to)
    {
        _from = from;
        _to = to;
    }

    public override bool Equals(object? obj)
    {
        obj = obj ?? throw new ArgumentNullException(nameof(obj));
        if (obj.GetType() != typeof(CurrencyPair)) return false;
        CurrencyPair pair = (CurrencyPair)obj;
        return _from.Equals(pair._from) && _to.Equals(pair._to);
    }

    // The book uses zero for a hashCode. Objects.hash() was only added to Java in Java 7. 
    // C# has HashCode.Combine() which is the equivalent of Objects.hash() in Java.
    public override int GetHashCode()
    {
        return HashCode.Combine(_from, _to);
    }
}